Feat/75 - Commitlint command

GeorgeCadwallader opened the pull request:
## Summary

Adds the commitlint command, replicating functionality from `1.x`.

Ref #75

Commit:
feat(commitlint): add commitlint packages

This will be used to lint the incoming commit messages.

Ref: #75

Commit:
feat(core): source control commands for retrieving commits

These two commands will be used to retrieve commits from the source
control provider. One retrieves the latest, current commit; and the
other returns a list of commits. The commit list can accept from and to
arguments that filter the list of commits that will be returned.

Ref: #75

Commit:
feat(commitlint): commitlint command

Command to lint commits using Commitlint. The command returns an output
of information about each commit if there are errors or warnings to
report about it.

Available arguments are the level of severity to lint, and the commit
hash or reference to start linting from.

Ref: #75

Commit:
chore: commit source con provider fixup

Commit:
test(commitlint): command assertions

Testing the commitlint command to make sure it's working as expected.
Tests include, validating correct and incorrect commit messages and log
output assertion.

Ref: #75

AdeAttwood commented on src/commands/commitlint.ts:
```diff
@@ -0,0 +1,141 @@
+  const commits = await sourceControl.getCommits(from as string, 'HEAD');
+
+  let hasErrors = false;
+  for (const commit of commits) {
```
I think this is OK for now. I would be nice to clean this impl up. I get its moved over, you know when you look at your old code and wonder what you were thinking :)

GeorgeCadwallader replied:
Yeah I get you :sweat_smile:

Yeah my plan was to just get it all working like before and then refactor later.

I'll create a ticket to tidy this up.

AdeAttwood commented on src/lib/source-control/git.ts:
```diff
@@ -17,6 +18,35 @@ const git: SourceControlProvider = {
+  },
+  getCommits: async (from, to) => {
+    const range = from && to ? `${from}..${to}` : '';
+    const { stdout: commitsString } = await run(`git log ${range} --pretty=format:"%s"`);
+    return commitsString.split('\n');
```
I don't think this is what we want. This will only get the subject, I think we should get the whole body so we can add validations on trailers and other things. Because the messages have new lines in them, the trick is to use zero bytes

```bash
git log --pretty=format:"%B%x00"
```

We can then split on `\0` to get each message in a array.

GeorgeCadwallader replied:
I get you. I'll apply your recommendation :+1:

AdeAttwood reviewed (changes requested):
Looks OK. I think we just need to add an impl that will throw an exception for sapling. This will get us up and running with git.

GeorgeCadwallader reviewed (commented):
I've got some free time tomorrow evening so will apply the amends then:

1. Create issue about tidying up code after it being ported from `1.x`
2. Replace command to get commits with git to use `git log --pretty=format:"%B%x00"`
3. Throw not supported related exceptions for sapling source control provider

Commit:
style: eslint fixes

Miscellaneous fixes to satisfy eslint.

Commit:
fix: throw not supported error for sapling methods

These methods are required by the source control provider but are not
supported with the sapling source control option.
